lambda-buffers-frontend/src/LambdaBuffers/Frontend/FrontM.hs

```diff
@@ -0,0 +1,238 @@
+  deriving stock (Eq, Show)
+
+type Symbol = TyRef ()
+type Scope = Map Symbol (ModuleName SourceInfo)
+
+data FrontErr
+  = ModuleNotFound (ModuleName SourceInfo) (Import SourceInfo) [FilePath]
+  | MultipleModulesFound (ModuleName SourceInfo) (Import SourceInfo) [FilePath]
+  | ImportCycleFound (ModuleName SourceInfo) (Import SourceInfo) [ModuleName ()]
+  | ModuleParseError FilePath ParseError
+  | ImportedNotFound (ModuleName SourceInfo) (ModuleName SourceInfo) (TyName SourceInfo) (Set (TyName SourceInfo))
+  | InvalidModuleFilepath (ModuleName SourceInfo) FilePath FilePath
+  | SymbolAlreadyImported (ModuleName SourceInfo) (Import SourceInfo) Symbol (ModuleName SourceInfo)
+  | TyRefNotFound (ModuleName SourceInfo) (TyRef SourceInfo) Scope
+  | DuplicateTyDef (ModuleName SourceInfo) (TyDef SourceInfo)
+  | TyDefNameConflict (ModuleName SourceInfo) (TyDef SourceInfo) (ModuleName SourceInfo)
+  deriving stock (Eq)
+
+showOneLine :: Doc a -> String
+showOneLine d = (renderShowS . layoutPretty (defaultLayoutOptions {layoutPageWidth = Unbounded}) $ d) ""
+
+instance Show FrontErr where
+  show (ModuleNotFound _cm imp impPaths) = showOneLine $ pretty (importInfo imp) <+> "Module" <+> pretty (importModuleName imp) <+> "not found in available import paths" <+> pretty impPaths
+  show (MultipleModulesFound _cm imp conflictingPaths) = showOneLine $ pretty (importInfo imp) <+> "Module" <+> pretty (importModuleName imp) <+> "found in multiple files" <+> pretty conflictingPaths
+  show (ImportCycleFound _cm imp visited) = showOneLine $ pretty (importInfo imp) <+> "Tried to load module" <+> pretty (importModuleName imp) <+> "which constitutes a cycle" <+> pretty visited
+  show (ModuleParseError _fp err) = showOneLine $ pretty err
+  show (ImportedNotFound _cm mn tn@(TyName _ info) available) = showOneLine $ pretty info <+> "Type" <+> pretty tn <+> "not found in module" <+> pretty mn <> ", did you mean one of" <+> pretty (Set.toList available)
+  show (InvalidModuleFilepath mn@(ModuleName _ info) gotModFp wantedFpSuffix) = showOneLine $ pretty info <+> "File name" <+> pretty gotModFp <+> "doesn't match module name" <+> pretty mn <+> "expected" <+> pretty wantedFpSuffix
+  show (SymbolAlreadyImported _cm imp sym alreadyInModuleName) = showOneLine $ pretty (importInfo imp) <+> "Symbol" <+> pretty sym <+> "already imported from module" <+> pretty alreadyInModuleName
+  show (TyRefNotFound _cm tyR@(TyRef _ _ info) scope) = showOneLine $ pretty info <+> "Type " <> pretty tyR <+> "not found in the module's scope" <+> (pretty . Map.keys $ scope)
+  show (DuplicateTyDef _cm tyDef) = showOneLine $ pretty (tyDefInfo tyDef) <+> "Duplicate type definition with the name" <+> pretty (tyName tyDef)
+  show (TyDefNameConflict _cm tyDef imn) = showOneLine $ pretty (tyDefInfo tyDef) <+> "Type name" <+> pretty (tyName tyDef) <+> "conflicts with an imported type name from module" <+> pretty imn
+
+type FrontM = ReaderT FrontRead (StateT FrontState (ExceptT FrontErr IO))
```

Comment:
To make it more flexible - I recommend we use 
```suggestion
type FrontEndT m a = MonadIO m => ReaderT FrontRead (StateT FrontState (ExceptT FrontErr m)) a
type FrontEnd a = FrontEndT IO a
```

In anycase - I'd drop the `M` - looks nicer.

Reply:
Done!